### ISSUE TITLE
Dx Station sends full exchange after user sends partial callsign

### ISSUE THREAD

**Issue by w7sst** [w7sst/MorseRunner#382]
## Description
Tim, N0UI, [reported this issue on groups.io](https://groups.io/g/MorseRunnerCE/topic/behaviors_in_1_85_1_85_1_for/109840974)...

> Mike:  
>  
> I noticed a behavior in the recent version of MorseRunnerCE I wanted to mention.  I have used Alex's Ver. 1.68 for a long time.  I was glad to use your Ver. 1.84 to prepare for the IARU this summer.  Adding contests has been appreciated.  Thanks for taking the time and making the effort to improve on this project.
>  
> When I upgraded to 1.85, and then 1.85.1 from 1.84, I noticed a behavior change in the IARU contest.  I send a partial call in response to a pileup, and A LOT of the stations will not just send their call, but their full exchange.  Now, I know you want this behavior occasionally, but the percentage seems to be around 90%.  I haven't tested this beyond the IARU contest setup yet.
>  
> I usually run quick 10 minute sessions, and it always seems to start up within the first couple of minutes, then keep occurring for the rest of the session.
>  
> Tim Raymer
> 73 de N0UI

My response...

> Hi Tim,
> 
> Thank you for reporting this issue. We did make some changes in how responses are generated for v1.85, so I appreciate the feedback.
> 
> Question... Are you sending your partial callsign using F5 key or the Enter key? The program's logic will treat each key differently. F5 only sends the partial call, whereas the Enter key will send your call and your exchange. Based on what you have said, I think you are using the F5 key only.
> 
> Thank you for your feedback.
> 
> 73, Mike W7SST

Sean, N3RTW, added...
> I haven't tracked the versions as closely, but in the latest MorseRunner CE it does this same thing on the CWTs. I use the F5 key to send a partial call and I'll usually get the full call and exchange in response.
> 
> Sean N3RTW

And finally Tim's response to my question...
>> Question... Are you sending your partial callsign using F5 key or the Enter key? The program's logic will treat each key differently. F5 only sends the partial call, whereas the Enter key will send your call and your exchange. Based on what you have said, I think you are using the F5 key only.
> 
> Mike,
> 
> Indeed, this is in an F5 Partial call scenario that this happens.
> 
> Thanks for taking note of the issue.
> 
> Tim Raymer
> 73 de N0UI

## Version
- 1.85 and 1.85.1

**Comment by w7sst** [comment]
My response back to groups.io...

Hi Tim, Thank you for clarifying. I found the code that is causing the problem. It is happening about 50% of the time - sometimes with the caller is sending the corrected call being sent twice and sometimes only once along with the full exchange.

I also notice that the calling station should only send their exchange when the user sends his exchange first. I realize the bug is the calling station is sending their exchange before the user has sent their exchange. In this reported case, the F5 key is sending the partial call without an exchange. Thus, the caller should only send their corrected call without an exchange.

This behavior change (in 1.85) was motivated by the following article: http://www.k1usn.com/Do%27s%20and%20Don%27ts%20and%20More.pdf

To quote this article (Italics added):

> **Am I My Brother’s Keeper?**
> 
> "I know of only one contest where you may be penalized if someone busts your call. In all others, the operator who logs the wrong call gets busted. As an S&Per, you could move up and down the band and work a bunch of runners, and if they screw up your call, you figure that’s their problem, not yours. But, trust me, when people start looking at results and they see that a lot of people busted K6RB’s call, it raises questions. You can’t prevent people from busting your call completely but you can help avoid it, somewhat, by paying attention to what they send. If you are K3PP and you hear a runner come back to K4PP,cdon’t just send “tu 5nn pa.” Send “K3PP (and maybe PP right after) then 5nn pa. You’ve done your job by helping the runner avoid a busted call.
> 
> Some people will hold off sending an exchange if a runner sends the wrong call. For example, if I sent K4PP 5nn ca to K3PP, he may send K3PP K3PP and pause without sending an exchange. I would then send K3PP 5nn ca. Then, he would send tu 5nn pa. However, doing this delays both the runner and pouncer, and only the runner is at risk of losing points. So, why not be a good citizen and send K3PP 5nn pa rather than holding off the exchange. If the runner gets it, and fixes it, then great. If not, you’ve done your job without delaying either of you."

I will study this further and either find a fix or reduce the percentage. Perhaps 50% is too much.

73 Mike W7SST

**Comment by w7sst** [comment]
This issue was fixed with merge #384.

- restores pre-1.85 behavior where DxStation sends only corrected callsign after user sends a partial callsign using F5 Key.
- v1.85/v1.85.1 was sending both corrected callsign and exchange. Only the corrected callsign should be sent. This is now fixed.
- The user can use the F5 Key to send a partial callsign.
- This fix restores original behavior where the DxStation will respond with callsign only after receiving an incorrect or partial callsign.
- If the user uses the Enter Key and sends both callsign and exchange, then the Dx Station will respond with their callsign and exchange. This will occur most of the time (~80%); otherwise the Dx Station will respond with only their corrected callsign.

**Comment by w7sst** [comment]
Laurent reported Issue #399 which I think was caused by the changes in this Issue. When validating this issue, make sure we consider the issue raised in 399.

**Comment by w7sst** [comment]
I sent an email to the original submitter, Tim, on groups.io and asked if could help validate this for us.

**Comment by w7sst** [comment]
Here is Tim's, N0UI, response via email:
> Mike:
> 
> I wanted to report back.  I can still duplicate the F7 not sending a question mark behavior.
> 
> Enter a partial incorrect call in the Call box.  Hit F5 to send the response.  After no response (due to the partial call being wrong), hit F7 to send a question mark.  No response from any station in the Pile-Up.  Clear the Call box, hit F7 again, normal responses from the Pile-Up. 
> 
> (As an aside, I noticed the timing of the F7 response seems to mimic the initial Pile-Up timing for a given sequence.  Would be more realistic if you used the variable timing like when you wait and not respond to a Pile-Up, then they send their callsigns again with different timing/sequencing.)
> 
> Second issue (Station sending exchange without being sent his complete callsign) happened once or twice in the sequences I sent, but it was much less prominent than with 1.85 and 1.85.1.  It seems to be at a realistic level that lends itself to the normal operating variability that you encounter in the real world.
> 
> Only other thing I noticed is that the QRN SEEMS louder relative to the sent callsigns in this version.  I always have QRN on (40/80 at night on Field Day, anyone?) so it caught my ear.  Just wanted to mention it, and highlight that it is an “it SEEMS” thing. 
> 
> I appreciate the opportunity to test the software for you.  If I can do anything else to verify, I will be glad to do so.  I will be out of pocket travelling this evening and tomorrow.  I can do more testing if you need it on Sunday and beyond.
> 
> Tim Raymer
> 73 de N0UI

It seems Tim is bring up a new issue here (F7 not sending a question mark behavior). I had not heard of this issue before. I will continue a conversation with Tim to better understand this issue.

**Comment by w7sst** [comment]
Here is the testing response from Sean, N3RTW...

> Hi Mike, I ran through a cycle of the CWT and it did it a couple of times, but not every time like it did before, which I think is intentional.
>
> My tests were to get a partial call and use that with F5 and see if I got a call back, or a call + exchange. Most of the time I got just the call. Twice out of around 20 contacts I got the call and exchange. I tried to test with mistakes in the partial and without. I'd like to say the mistakes were intentional but I can't :P
>
> Thanks for your continued work on Morse Runner CE
>
> 73 de  Sean N3RTW
